app/models/account/gateway.rb

```diff
@@ -14,12 +14,10 @@ module Account::Gateway
   end
 
   # Payment gateway this account accepts payments through.
-  def payment_gateway
-    @payment_gateway ||=
-      if payment_gateway_type.present?
-        gateway_class = PaymentGateway.implementation(payment_gateway_type)
-        gateway_class.new(payment_gateway_options || {})
-      end
+  def payment_gateway(**options)
+    return if payment_gateway_type.blank?
+
+    PaymentGateway.implementation(payment_gateway_type, options).new(payment_gateway_options || {})
```

Comment:
No longer memoized?

Reply:
No need imo 🤷🏼‍♀️

Reply:
But I wouldn't mind either

Reply:
@guicassolato Think that now it depends on 'options' and if we make 2 calls to this method with different options, we should receive 2 different responses and not the prior memoized one that would no longer be correct for the 2nd call

Reply:
The scenario I've mentioned is tested 👍🏼

Reply:
I gotta check that this object does not need to be the same every time that it is called or do something about it 👍🏼

Reply:
I think that it is not a problem, but I will expose here why, just in case.

To know from where it is used: `ag '(\s|\.)(provider_)?payment_gateway(?!_)'`
![image](https://user-images.githubusercontent.com/11318903/104158294-4ed70100-53ed-11eb-8cb6-f6240934fee1.png)
I will skip explanation of files inside: `test/` `features/` `doc/`

1. `app/models/account/billing.rb` is inside the method edited in this PR... all good for sure here 👍🏼
2.  `app/models/account/credit_card.rb` called in 3 places inside different methods, but the results shouldn't change for being a different object:
https://github.com/3scale/porta/blob/342c3f4860454215d8d5ef057b10e5352606ef67/app/models/account/credit_card.rb#L103
https://github.com/3scale/porta/blob/342c3f4860454215d8d5ef057b10e5352606ef67/app/models/account/credit_card.rb#L139
https://github.com/3scale/porta/blob/342c3f4860454215d8d5ef057b10e5352606ef67/app/models/account/credit_card.rb#L147
3. `app/models/account/gateway.rb` is this same file where we are commenting about this... no problem here for sure 👍🏼 
4. `app/lib/payment_gateways/payment_gateway_crypt.rb` just to log the display_name. It shouldn't be a problem.
https://github.com/3scale/porta/blob/342c3f4860454215d8d5ef057b10e5352606ef67/app/lib/payment_gateways/payment_gateway_crypt.rb#L46-L55
5. `app/lib/payment_gateways/authorize_net_cim_crypt.rb` used to create or get a customer profile
https://github.com/3scale/porta/blob/342c3f4860454215d8d5ef057b10e5352606ef67/app/lib/payment_gateways/authorize_net_cim_crypt.rb#L28-L33
https://github.com/3scale/porta/blob/342c3f4860454215d8d5ef057b10e5352606ef67/app/lib/payment_gateways/authorize_net_cim_crypt.rb#L123-L130
6. `app/views/finance/provider/settings/show.html.slim` shouldn't be a problem because it is getting them from a variable.
https://github.com/3scale/porta/blob/342c3f4860454215d8d5ef057b10e5352606ef67/app/views/finance/provider/settings/show.html.slim#L72-L86
7. `lib/tasks/credit_cards.rake` creates a customer profile, but it is using the object only once so no problem
https://github.com/3scale/porta/blob/342c3f4860454215d8d5ef057b10e5352606ef67/lib/tasks/credit_cards.rake#L10
8. `lib/developer_portal/app/controllers/developer_portal/admin/account/authorize_net_controller.rb` ... There are 2 actions, 'show' and 'hosted_success', for 'show' I am sure that it is not a problem because it only uses the object to fetch the options, and the options are always the same no matter how many different objects there are... for 'hosted_success' I am not so sure because it calls twice to `cim_gateway` and I am not sure if this can be a different object, but if it can't be, then this particular case is easy to fix:
https://github.com/3scale/porta/blob/342c3f4860454215d8d5ef057b10e5352606ef67/lib/developer_portal/app/controllers/developer_portal/admin/account/authorize_net_controller.rb#L13-L35
9. `lib/developer_portal/app/views/developer_portal/accounts/payment_gateways/edit.html.liquid` for sure not a problem because it is only checking the type fo payment_gateway and that never changes.
https://github.com/3scale/porta/blob/342c3f4860454215d8d5ef057b10e5352606ef67/lib/developer_portal/app/views/developer_portal/accounts/payment_gateways/edit.html.liquid#L23
10. `lib/developer_portal/app/views/developer_portal/accounts/payment_gateways/show.html.liquid` same as before, only checking the type 😉 
11. `lib/developer_portal/lib/liquid/drops/provider.rb` just a false positive in the search apparently 😄 
https://github.com/3scale/porta/blob/342c3f4860454215d8d5ef057b10e5352606ef67/lib/developer_portal/lib/liquid/drops/provider.rb#L50-L53

Unfortunately, this search is not exhaustive enough... but all the other cases that come to mind do not give new results of stuff that I have not checked:
![image](https://user-images.githubusercontent.com/11318903/104160551-9eb7c700-53f1-11eb-884a-ba7b8a9ea2eb.png)

Reply:
The only one that I was doubting about, file number 8, fixed in #2329